network/src/peer/mod.rs

```diff
@@ -409,26 +498,36 @@ where
 }
 
 pub struct PeerHandle {
-    peer_id: PeerId,
+    connection_metadata: ConnectionMetadata,
+    network_context: Arc<NetworkContext>,
     sender: channel::Sender<PeerRequest>,
 }
 
 impl Clone for PeerHandle {
     fn clone(&self) -> Self {
         Self {
-            peer_id: self.peer_id,
+            network_context: self.network_context.clone(),
+            connection_metadata: self.connection_metadata.clone(),
             sender: self.sender.clone(),
         }
     }
 }
 
 impl PeerHandle {
-    pub fn new(peer_id: PeerId, sender: channel::Sender<PeerRequest>) -> Self {
-        Self { peer_id, sender }
+    pub fn new(
+        network_context: Arc<NetworkContext>,
+        connection_metadata: ConnectionMetadata,
```

Comment:
Maybe we should `Arc<ConnectionMetadata>`... since we have the same struct copied in 4 places for each peer that connects. It also contains a `NetworkAddress`, which can be ~140 bytes since lots of validators use their long LB dns names. I want to make sure we keep our memory usage in check so we don't like alloc 1 MiB per peer and then suddenly only support 1000 connections before OOM lol

Reply:
Let's open an issue for it, and do in a separate PR